### ISSUE TITLE
characters in neovim are behaving 'weird'

### ISSUE THREAD

**Issue by yaghc** [kovidgoyal/kitty#308]
I am using neovim with mainly LanguageClient-neovim ~~and nvim completion-manager~~, but I can't use error checking in kitty. I am creating an issue here, not in any of those two plug-ins, since this bug does not occur in alacritty nor asciinema.
You should see, what I mean by this here: https://streamable.com/smvkk
If you need my kitty or neovim config, just ask, even though the same bug occurred when using (almost) default configs.
If it is not caused by kitty, sorry then.
I am using kitty-git v0.7.1.r14.g561e63e-1 from the AUR.

**Comment by kovidgoyal** [comment]
Since I am not familiar with any of the software you mention, I have no idea what I am supposed to be looking for. I can tell you that I use YouCompleteMe in vim in kitty all day long with no strangeness of any kind. Where in that video is the strangeness and what is the strangeness?

**Comment by yaghc** [comment]
Look at around 0:24, the characters are moving rather weird at this point (the video was perhaps a bit too long).
When doing the same thing later in alacritty nothing like that happens.
A minimal config file to reproduce this issue: https://hastebin.com/iyapevebup.vim
You might need to install the rust language server (https://github.com/rust-lang-nursery/rls), if you want to try this yourself.
The same thing should happen when using another language server that supports errors, warnings or tips.

**Comment by kovidgoyal** [comment]
I'm guessing this is because kitty and neovim are disagreeing about the width of the special characters. For example, in kitty u+2716 is supposed to be two characters wide. And this correct, because in the unicode standard, u+2716 is classified as an emoji, see http://www.unicode.org/Public/emoji/5.0/emoji-data.txt

You can check that kitty treats it as two characters wide by running:
```
python3 -c "print('\u2716\naa')"
```

And in neovim it is treated as a single width character. You can test this by typing ```<C-v>u2716``` and see that neovim thinks the cursor has moved one cell forward only.

I though neovim generated its width tables from the unicode standard, but perhaps I am misremembering. Maybe @justinmk can comment.

**Comment by kovidgoyal** [comment]
You can workaround it by using `use_system_wcwidth=yes` IIRC the system wcwidth function treats that character as width 1.

**Comment by kovidgoyal** [comment]
Yes, I see that neovim for some reason excludes characters below 0x1f000 from being considered as emoji: https://github.com/neovim/neovim/blob/master/src/nvim/generators/gen_unicode_tables.lua#L251

You will have to ask its developers why that is. IMO all applications should stick to the unicode standard, otherwise there will end up being hopeless confusion.

**Comment by kovidgoyal** [comment]
@egmontkob you might be interested in this issue as well.

**Comment by egmontkob** [comment]
Thanks for the heads-up. My understanding is that this emoji stuff doesn't suddenly override wcwidth, only if followed by an emoji presentation selector. See [VTE bug 767529](https://bugzilla.gnome.org/show_bug.cgi?id=767529). U+2716 is single-width in xterm, urxvt, pterm, st, konsole, terminology and vte.

**Comment by kovidgoyal** [comment]
I dont think that's right. Testing in gnome-terminal

```
python3 -c "print(chr(0x1f9d0))"
```

Displays a wide character. Despite not being followed the  presentation selector. 

Maybe you meant your comment to apply to only emoji characters that do not default to emoji_presentation? If so, I can agree with that. I should probably change kitty to behave that way as well.

**Comment by egmontkob** [comment]
> Maybe you meant your comment to apply to only [...]

Yup, most likely. I'm really not familiar with this topic in details.

**Comment by yaghc** [comment]
Thanks for your answers!

**Comment by kovidgoyal** [comment]
I have pushed code to the emoji branch to change kitty to only assume chars with the emoji presentation style are wide. This should fix this issue. I need to add support for the unicode emoji presentation variant selectors as well, once that is done, I will merge into master.
